src/renderers/widget.ts

```diff
@@ -0,0 +1,354 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+
+import {
+  ansi_to_html, escape_for_html
+} from 'ansi_up';
+
+import * as CodeMirror
+  from 'codemirror';
+
+import 'codemirror/addon/runmode/runmode';
+
+import {
+  requireMode
+} from '../codemirror';
+
+import * as marked
+  from 'marked';
+
+import {
+  RenderMime
+} from '../rendermime';
+
+import {
+  Message
+} from 'phosphor/lib/core/messaging';
+
+import {
+  Widget
+} from 'phosphor/lib/ui/widget';
+
+import {
+  typeset, removeMath, replaceMath
+} from './latex';
+
+
+/*
+ * The class name added to common rendered HTML. 
+ */
+const HTML_COMMON_CLASS = 'jp-RenderedHTMLCommon';
+
+/*
+ * The class name added to rendered HTML. 
+ */
+const HTML_CLASS = 'jp-RenderedHTML';
+
+/*
+ * The class name added to rendered markdown. 
+ */
+const MARKDOWN_CLASS = 'jp-RenderedMarkdown';
+
+/*
+ * The class name added to rendered Latex. 
+ */
+const LATEX_CLASS = 'jp-RenderedLatex';
+
+/*
+ * The class name added to rendered images. 
+ */
+const IMAGE_CLASS = 'jp-RenderedImage';
+
+/*
+ * The class name added to rendered text. 
+ */
+const TEXT_CLASS = 'jp-RenderedText';
+
+/*
+ * The class name added to rendered javascript. 
+ */
+const JAVASCRIPT_CLASS = 'jp-RenderedJavascript';
+
+/*
+ * The class name added to rendered SVG. 
+ */
+const SVG_CLASS = 'jp-RenderedSVG';
+
+/*
+ * The class name added to rendered PDF. 
+ */
+const PDF_CLASS = 'jp-RenderedPDF';
+
+
+// Support GitHub flavored Markdown, leave sanitizing to external library.
+marked.setOptions({
+  gfm: true,
+  sanitize: false,
+  breaks: true,
+  langPrefix: 'cm-s-default language-',
+  highlight: (code, lang, callback) => {
+    if (!lang) {
+        // no language, no highlight
+        if (callback) {
+            callback(null, code);
+            return;
+        } else {
+            return code;
+        }
+    }
+    requireMode(lang).then(spec => {
+      let el = document.createElement('div');
+      if (!spec) {
+          console.log(`No CodeMirror mode: ${lang}`);
+          callback(null, code);
+          return;
+      }
+      try {
+        CodeMirror.runMode(code, spec.mime, el);
+        callback(null, el.innerHTML);
+      } catch (err) {
+        console.log(`Failed to highlight ${lang} code`, err);
+        callback(err, code);
+      }
+    }).catch(err => {
+      console.log(`No CodeMirror mode: ${lang}`);
+      console.log(`Require CodeMirror mode error: ${err}`);
+      callback(null, code);
+    });
+  }
+});
+
+
+/*
+ * A widget for displaying any widget whoes representation is rendered HTML
+ * */
+export
+class RenderedHTMLCommon extends Widget {
```

Comment:
Yes, otherwise the exported classes that use the base class complain in the TS compiler